### PR TITLE
fix: version does not git-commit nor git-tag when package in subdir

### LINKED ISSUE #2010: [BUG] npm-version does not git-commit nor git-tag when package in subdirectory
### Current Behavior:

`npm version <version>` is not committing the modified package.json or package-lock.json; nor git-tagging.

We were using `npm version {major|minor|patch}` extensively (and successfully). It stopped working once the package was moved out of the root of the repo and into a subdirectory.

### Expected Behavior:
`npm version <version>` should continue to create a git-commit and git-tag as indicated in the [docs](https://docs.npmjs.com/cli-commands/version.html):

> If run in a git repo, it will also create a version commit and tag. This behavior is controlled by git-tag-version (see below), and can be disabled on the command line by running npm --no-git-tag-version version. It will fail if the working directory is not clean, unless the -f or --force flag is set. 

### Steps To Reproduce:
1. initialize an npm package in the root of an initialized git repo
2. `npm version minor` successfully bumps the version, commits and tags
3. move the npm package into a subdirectory of the repo
4. `npm version minor` still bumps the version in package.json and package-lock.json, but git is not committed nor tagged.

### Environment:
- macOS 10.15.7
- node: v14.13.1
- npm: 6.14.8
- git: git version 2.28.0

This is apparently an existing bug going as far back as npm v3: https://github.com/npm/npm/issues/18795

### PR DESCRIPTION
**Pull request by timothybonci**
Change the git detection to use git.find in order to walk up through directories to find the git repo.

Add use case for detecting if you're in a subdirectory that is a workspace, and make that case the exception for creating the commit and tag.

## References
Fixes #2010

### CODE REVIEW

**Comment by lukekarrys** [comment]
Thanks for the new PR for this @timothybonci! It looks good so far, but CI is showing a few more tests in `test/lib/commands/version.js` that are failing now. Would you be able to look at those too?

**Review by lukekarrys — CHANGES_REQUESTED** [review]
libnpmversion tests are passing but cli tests for `commands/version` are failing.

**Comment by timothybonci** [comment]
Yep, I missed that test locally.  Thanks for kicking off the workflows, and thanks for fixing the git config in the .github ci.  I'll fire up the debugger.

**Comment by timothybonci** [comment]
Aha, there was a bug in the internal caller in lib/commands/version.js where we were passing the flag instead of the option property, so we were never properly overriding gitTagVersion with false.  These tests were passing before because we'd short circuit the doGit check on git.is returning false.  Now that we use git.find and find the repo, we pass to the next check for doGit, which properly fails.  Now that it's fixed, we short circuit on the very first check for gitTagVersion, and correctly don't attempt any git operations.

Long story short, I think I fixed it.  I'd appreciate another run of the workflows.

**Comment by timothybonci** [comment]
@lukekarrys would you please approve the test workflows to test the latest changes?

**Comment by timothybonci** [comment]
Rebased to latest and fixed the two outstanding lint issues.

**Comment by timothybonci** [comment]
Rebased to latest again.  @lukekarrys could you review?

**Comment by lukekarrys** [comment]
Sorry for the delay in reviewing this @timothybonci. I am out for a few weeks and will review it when I get back. Thanks for your continued work on this!

**Comment by 1oglop1** [comment]
@lukekarrys are there any updates on this? It has been quite a few weeks. As a newcomer to NPM, I wasted a lot of time trying to figure out why this feature is not working according to docs :(.

**Comment by martinmiglio** [comment]
bumping because I _need_ this feature.

**Comment by markhougaard** [comment]
This would be extremely helpful to get merged. I've wasted a few days trying to understand why this behavior is different from the official docs. Workarounds are nice, but it's better to get it fixed properly.

**Comment by wraithgar** [comment]
~~Typically we try to avoid unit tests like those written in `test/is-workspace-safe.js`~~

~~Our testing approach is that we need to be able to reach it all from the surface area of the module itself.  Unit testing in npm has historically led to a lot of dead code and unneeded complexity handling cases that were impossible in whole.~~

~~It's fine to test those things in a separate file for organization but they need to reach those lines of code through `libnpmversion` itself, not by requiring the module directly.~~

ETA: it looks like this repo is still all unit tests.  We're not asking you to implement this so the unit test will have to do for now

**Comment by Stehsegler** [comment]
We're also urgently waiting for this bugfix.

@lukekarrys @wraithgar - If I understand the latest comment correctly, the PR is now ready to be merged? If not, what still needs to be done?

Thank you all for your engagement in this product!

**Comment by dmcweeney** [comment]
@lukekarrys @wraithgar @timothybonci Wondering if there will be an progress made on this PR for this bug fix?

**Comment by wraithgar** [comment]
I think what's left is to fix the merge conflicts and to make sure we're ok with the tag it will make.  Workspaces need to be tagged differently than the root project because without any changes they will collide with each other and the root project (since the tag is only the version, it's not namespaced in any way)

**Comment by timothybonci** [comment]
After a year of silence, I'm not going to contribute further.  I'm relinquishing any possible claims I may have on anything I've written in these commits/MR.  I wish you all the best of luck.
